compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp

```diff
@@ -1179,6 +1115,262 @@ FailureOr<Value> DataTiledMMAAttr::buildMmaOperation(OpBuilder &builder,
   return acc;
 }
 
+//===----------------------------------------------------------------------===//
+// VirtualMMA Attributes
+//===----------------------------------------------------------------------===//
+
+VirtualMMAAttr VirtualMMAAttr::get(MLIRContext *context,
+                                   VirtualMMAIntrinsic type) {
+  auto intrinsicAttr = VirtualMMAIntrinsicAttr::get(context, type);
+  return VirtualMMAAttr::get(context, intrinsicAttr);
+}
+
+static OpaqueMmaLayout getOpaqueVMMALayout(MLIRContext *context,
+                                           VirtualMMAIntrinsic type) {
+  Type f8E4M3FNUZ = Float8E4M3FNUZType::get(context);
+  Type f16 = Float16Type::get(context);
+  Type f32 = Float32Type::get(context);
+
+  switch (type) {
+  case VirtualMMAIntrinsic::VMFMA_F32_16x16x32_F8E4M3FNUZ: {
+    return OpaqueMmaLayout{16, 16, 32, f8E4M3FNUZ, f8E4M3FNUZ, f32};
+  }
+  // V(Virtual)MFMA instructions which have 2 mfma instructions interleaved
+  // along the k dimension.
+  case VirtualMMAIntrinsic::VMFMA_F32_16x16x32_F16: {
+    return OpaqueMmaLayout{16, 16, 32, f16, f16, f32};
+  }
+  case VirtualMMAIntrinsic::VMFMA_F32_32x32x16_F16: {
+    return OpaqueMmaLayout{32, 32, 16, f16, f16, f32};
+  }
+  }
+  llvm_unreachable("unhandled mfma layout type");
+  return OpaqueMmaLayout{};
+}
+
+std::tuple<Type, Type, Type> VirtualMMAAttr::getABCElementTypes() const {
+  MLIRContext *ctx = getContext();
+  auto opaqueLayout = getOpaqueVMMALayout(ctx, getIntrinsic().getValue());
+  return {opaqueLayout.aType, opaqueLayout.bType, opaqueLayout.cType};
+}
+
+std::tuple<VectorType, VectorType, VectorType>
+VirtualMMAAttr::getABCVectorTypes() const {
+  // Check https://github.com/ROCm/amd_matrix_instruction_calculator for
+  // instruction details. Note here we are returning the number elements, while
+  // amd_matrix_instruction_calculator tells us about the number of 32-bit
+  // registers. So need to adjust accordingly. All vectors should be 1-D.
+  auto [A, B, C] = getABCElementTypes();
+  switch (getIntrinsic().getValue()) {
+  case VirtualMMAIntrinsic::VMFMA_F32_16x16x32_F8E4M3FNUZ:
+  case VirtualMMAIntrinsic::VMFMA_F32_16x16x32_F16: {
+    auto aType = VectorType::get({8}, A);
+    auto bType = VectorType::get({8}, B);
+    auto cType = VectorType::get({4}, C);
+    return std::make_tuple(aType, bType, cType);
+  }
+  case VirtualMMAIntrinsic::VMFMA_F32_32x32x16_F16: {
+    auto aType = VectorType::get({8}, A);
+    auto bType = VectorType::get({8}, B);
+    auto cType = VectorType::get({16}, C);
+    return std::make_tuple(aType, bType, cType);
+  }
+  }
+  // This should not happen but just to make GCC happy.
+  return std::make_tuple(VectorType{}, VectorType{}, VectorType{});
```

Comment:
```suggestion
  return {};
```

Also add `llvm_ureachable` if it's not supposed to happen?

Reply:
@kuhar , `llvm_unreachable` is a powerful optimization tool, by allowing the compiler to assume a place to be unreachable, but it's also dangerous if it's ever reached, so it only makes sense to use it in very, very performance-critical code.  In all other cases, if you want to catch something that's not supposed to happen, maybe you want an assertion (`assert(false)`); all the usual objections against assertions ("don't risk crashing the compiler on bad user input") also apply even more strongly against `llvm_unreachable`.

Reply:
Also note that here, GCC is saying something very important and Clang's behavior is borderline buggy: these TableGen-generated enums unfortunately have a specified underlying type:

```
enum class MMAIntrinsic : uint32_t {
```

Unfortunately, the `: uint32_t` makes a big difference here: it makes it legal to form values that aren't any of the enumerator values. So when Clang says "you dont need a default: case in this switch because it would be UB anyway if the enum value were none of the enumerators", that's only correct for enums without a specified underlying type.

Here, unless you want to audit every caller, we have to treat it as a possibility that the enum value is none of the enumerators and so that "error" case is in fact reachable.

Reply:
The way I understand it, `llvm_unreachable` translated to:
* Debug ==> `assert(false)`
* Release ==> `assume(false)`

While plain `assert` disappears in Release. So to me `llvm_unreachable` is fine as a tool to stop the compiler from complaining about missing return values etc., when there's something else that structurally checks that all the cases were handled. And I think this will happen by warnings about unhandled `switch` cases.

But I don't care about it much in either direction

Reply:
I replied to the first comment before I saw your second one; assert makes sense to me then. I didn't realize it was an open set of possible values.

Reply:
> While plain assert disappears in Release.

Yes. Apparently the idea is that the same control paths are exercised in Debug and Release builds. That assumption seems untenable in practice. Users use Release bugs, and they will run into corner cases that weren't hit by tests and local developer invocations on debug builds.

Reply:
done!

Reply:
Filed https://github.com/llvm/llvm-project/issues/115345